doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+that does not result in termination. If the handling code itself fails (e.g. insufficient resources)
+no recovery is possible and the system terminates.
+
+### Non-recoverable Error Handling
+
+A non-recoverable error handling strategy will always call terminate after the handling code was executed.
+Handler code may be used to initiate a graceful shutdown. An example are (strict) asserts.
+
+## Requirements
+
+In the following we list the requirements of the Error Handling (EH) strategy. The term platform
+refers to the usage environment of the error handler. There can be various platforms,
+e.g. for regular operations, testing, different OS etc.
+
+### Functional
+
+1. The EH shall support error codes.
+    - configurable on a platform basis
+    - errors from different platforms shall not be usable (checked at compile time)
+    - uniqueness is not enforced (can be done by the layer above if desired)
+    - error codes shall be optional (if none is used, it is mapped to some default error code)
+1. The EH shall support multiple error levels.
+    - the individual error levels are defined by the platform
+    - there has to be at least a FATAL error level (most severe error level)
+    - the error levels shall be ordered from least severe to most severe
+    - if the code requires an ERROR level that is not supported by the platform it shall not compile
+1. The EH shall support a non-recoverable error handling strategy in case of FATAL errors.
+    - non-recoverable errors shall call terminate after optional handling and logging code has been executed
+1. The EH shall support a recoverable error handling strategies for non-FATAL errors.
+    - recoverable error handling shall support execution of sufficiently general user actons (callbacks)
+1. The detection mechanism of the EH shall allow logging the error.
+    - user defined-logging shall be supported
+    - there shall be a default implementaton for logging
+1. The EH shall allow to trace the error to its source location (file, function, line).
+1. The EH shall be disableable for non-FATAL errors.
+    - disabling shall be possible up to some level at compile time; this level shall be non-FATAL
+    - can be used if there is sufficient trust in the system to disable e.g. warning checks
+    - by default all levels defined by the platform are enabled
+    - disabling shall cause the affected error logic in the code to cause no runtime overhead
+
+### Technical Constraints
+
+1. The EH shall be uniquely defined for each platform.
+    - only one error handler can be active for a platform at any time
+    - the EH shall not be changed while the system runs
+    - the EH must be configurable at runtime in the init phase
+      (for tests, but it might be sufficient to change it at compile time if only required for tests)
+    - for tests it must be possible to override the FATAL error reaction causing termination
```

Comment:
No we don't need that. In test we could throw an exception in our `optional handling` code snippet which the error handler has to provide also for fatal errors.
But we need to get rid of `noexcept` for that to work.

Reply:
We should not always throw something in FATAL mode, because this would have impacts on the binary size (possible impacts on execution speed (mainly in the bad case where we not care too much about it, but this is not guaranteed).

Therefore we should leave the option open to override the FATAL behavior. It is not complicated in code nor does it cause additional costs.

Reply:
We could have a custom error handler only for the tests which throws the error code as exception.

Reply:
So you never run the tests with the same compiler flags as the production code?